lib/Registrator.js

```diff
@@ -142,11 +142,17 @@ Registrator.prototype = {
             expires = this.expires;
           }
 
-          // Re-Register before the expiration interval has elapsed.
+          // Re-Register or emit an event before the expiration interval has elapsed.
           // For that, decrease the expires value. ie: 3 seconds
           this.registrationTimer = setTimeout(function() {
             self.registrationTimer = null;
-            self.register();
+            // If there are no listeners for registrationExpiring, renew registration
+            // If there are listeners, let the function listening do the register call
+            if (self.ua.listeners('registrationExpiring').length === 0) {
+              self.register();
+            } else {
+              self.ua.emit('registrationEnding');
+            }
           }, (expires * 1000) - 3000);
```

Comment:
Could you please increase those 3000 ms to something higher such as 10000ms? 3 seconds is a bit risky if async operations would be performed before re-registering.